docs/RFCS/20230605_workload_level_index_recommendation.md

```diff
@@ -0,0 +1,348 @@
+indexes, what should we do? Along the idea of trie tree, we can use some internal node to build index, here is one 
+example:
+```
+    t
+    |
+    k
+  /   \
+i       j 
+```
+We can build a index `(k)` instead of two `(k, i)` and `(k, j)` but still get a good performance. The idea behind is 
+merging leaf nodes. As for the example above of two leaf nodes, we can find the 
+[lowest common ancestor (LCA)](https://en.wikipedia.org/wiki/Lowest_common_ancestor) to represent them, since we reduce 
+the number of indexes by 1 and keep the longest share between them. If we need to merge more than two indexes, we can 
+borrow the idea from [Huffman coding](https://en.wikipedia.org/wiki/Huffman_coding), each time we merge the most 
+expansive index with its nearest indexes (may from the leaf node or internal node) and then repeat this process until 
+the budget satisfied. Just like the example above, we merge the indexes `(k, i)` and `(k, j)` to `(k)`. If there is no 
+sharing between existing indexes, we can get rid of those expansive one. 
+
+In order to complete the algorithm above, we need to think of some objective function to estimate the cost for different 
+indexes based on the number of indexed columns, the number of storing columns, different types of columns, usage in the 
+whole workload, etc. 
+
+As mentioned in the finding index candidates, the frequency is an important aspect we should pay attention to, in 
+`system.statement_statistics`, there is a column called `statistics` containing the times of on a type of query has 
+been executed, we can use this to assign a weight for each node in the tree so that the merging algorithm will cover the 
+importance of the indexes (by frequency). If we collect from the raw sql statements mentioned in the future work, we can 
+re-compute the frequency from scratch and use such more precise information to assign the weight.
+
+## Drop index
+
+As for the `DROP INDEX`, since we select all the leaf nodes before, so we can directly drop them without hesitation 
+(covered by all the leaf nodes). Nevertheless, we only select some internal nodes to build indexes  when we need to 
+merge some leaf nodes for cost optimization so that not everything in the tree is covered, meaning that some dropped 
+indexes are not covered by the indexes selected to create. Since the space and time budget is confined, we can also take 
+dropping indexes into account, the cost function can be used to calculate the space released by dropping one index. 
+Following the idea of the trie tree, we can add all the indexes to drop to the tree. When we decide to merge two 
+indexes, there may exist some indexes to drop along the paths to their LCA, when we compute the space to save, we can 
+minus those space occupied by the indexes to drop since we can't cover them once we merge these two indexes. Thus, our 
+merging process will be more comprehensive.
+
+## Rationale and alternatives
+
+...
```

Comment:
It would be good to discuss the partial ordering solution from the paper that we read here. I think your approach (starting with the recommendations from single statements and merging them with a trie) sounds good as a first approach that is likely to get good results, but in the future we may want to consider using partial orders since that could get us closer to an optimal result. To merge partial orders, we could make use of the [OrderingChoice](https://github.com/cockroachdb/cockroach/blob/1a339b712842a768d48bf658fa00c3f5af7b02bb/pkg/sql/opt/props/ordering_choice.go#L61) data structure.